doc/developer/design/20240625_subsource_deprecation_blue_green_sources.md

```diff
@@ -0,0 +1,317 @@
+# Subsource deprecation & source blue-green schema changes
+
+This design encompases 2 existing proposals also known as
+'Subsource Demuxing' and 'Subsources as Tables'.
+
+Associated:
+- https://github.com/MaterializeInc/materialize/issues/20208
+- https://github.com/MaterializeInc/materialize/issues/24843
+- https://github.com/MaterializeInc/materialize/issues/15897
+- https://github.com/MaterializeInc/materialize/issues/17021
+
+## The Problems
+
+TLDR:
+- Subsources are confusing and not implemented uniformly across source types
+- It's impossible to do a blue-green workflow in sources for handling
+  schema-changes in upstream systems
+
+Users currently have to learn about *sources*, *subsources*, *tables*,
+*materialized views*, *indexes*, and *sinks* when onboarding to Materialize.
+
+If they'd like to ingest data from both an upstream Postgres/MySQL database
+and a Kafka topic, they also need to learn the fragmented user interface
+for dealing with data ingested from upstream data sources. Our existing
+model of a *source* does not provide a consistent interface among source types.
+
+Single-output Kafka sources put their data under the source's primary relation,
+whereas multi-output Postgres/MySQL sources do not allow querying the source's primary
+relation and instead contain *subsources*, an entirely separate concept.
+
+Our existing *source* models also make it difficult to deal with upstream schema changes,
+such that users need to drop and recreate a *source* or *subsource* and any dependent objects
+to begin ingesting data with a new schema. This can cause long-periods of downtime if the
+new source needs to reingest the entire upstream history. This can also cause them to
+lose their _reclocking_ information which makes it impossible to implement a correct
+end-to-end *source* -> *sink* pipeline.
+
+This change aims to simplify the user experience and create a more unified
+model for managing data ingested from an upstream source, while providing
+more flexibility in handling upstream schema changes.
+
+
+## Success Criteria
+
+- The mental overhead & learning time for users to ingest data from an upstream system
+  is decreased.
+- All source types use a consistent interface to retrieve progress information and
+  ingested data. The primary *source* object represents the same concept across
+  all source types.
+- The same upstream table/topic can be ingested into multiple collections in Materialize,
+  each with a potentially different projection of the upstream schema. This enables
+  a user to do blue-green schema-change workflows without unavailability.
```

Comment:
> We now want to allow users to not have to trigger a blue/green deployment to incorporate source schema changes downstream immediately.

I'm not following how that will be possible with this design? That seems like it's part of the goals of in-place schema changes, but this design is a precursor to that work. A user will still need to add a new table that uses the new upstream system schema, and then blue-green that table with their existing one to avoid updating downstream dependencies. Or am I missing something?

Reply:
I agree with @rjobanp's take, fwiw! Isn't the problem we're solving with subsources-as-tables that you _can't_ today use dbt to manage a blue/green deployment of a PostgreSQL or MySQL source?

Reply:
> A user will still need to add a new table that uses the new upstream system schema, and then blue-green that table with their existing one to avoid updating downstream dependencies.

Maybe we're looking at it from different perspectives. If you care about incorporating (potentially breaking) changes into a source and the downstream models are not owned by the same (or just one) team, you might want to run two versions of the same source with the different schemas for a while, then eventually deprecate the legacy one — which is now possible. You still need to run a blue/green deployment _eventually_! That's the workflow laid out in the [schema changes roadmap](https://docs.google.com/document/d/166qscfKs6AEP3HMZQuaT1NCx7ky4tO_E4jLnuApweKk/edit#bookmark=id.yo9ntbdzlc52).

> Isn't the problem we're solving with subsources-as-tables that you can't today use dbt to manage a blue/green deployment of a PostgreSQL or MySQL source?

You can't manage individual subsources in dbt (_i.e._ bring them into the dbt context as a standalone model), which will now be possible. For blue/green, since you can't swap individual objects, as long as the source tables are in the same cluster as the source, don't you necessarily need to recreate all source tables anyway? I'm not following what changes there!

Reply:
> The existing blue/green workflow already allows users to apply changes with no unavailability. 

I don't actually think this is the case unless you create a whole new top-level source that then re-ingests all your upstream data.

My take is that this _enables_ blue-green workflows without requiring a new top-level source, and it enables DBT workflows since each object is its own distinct thing.

The idea is that if your upstream postgres db adds column `foo` to table `bar`, and you already have a `bar` table in MZ linked source `baz`, you can now create a table called `bar2` that reads from source `baz` but with the new column `foo`. Then you can create new downstream objects based off of `bar2` and eventually blue-green the downstream object names with zero downtime.

Reply:
> I don't actually think this is the case unless you create a whole new top-level source that then re-ingests all your upstream data.

Exactly. But this is what happens today.

> My take is that this enables blue-green workflows without requiring a new top-level source,

And this is my understanding, too! Sorry — in the previous comment I was thinking about the scenario where source tables and their dependent objects are co-located (because you'd need to blue/green the whole cluster). We can disallow this scenario, like we do for sinks. All environments currently using blue/green deployments follow this patterns of separating data ingestion and data modeling.

Reply:
clarified on slack. The intent is that this allows source-tables to be part of a blue-green deployment directly, rather than having to be managed outside of it. It also avoids the need to reingest the upstream and lose reclocking information (though that is described in the next bullet point)